### PR TITLE
Bring back Orchard.OpenId module and migrate to OpenIddict beta2

### PR DESCRIPTION
**Pull request by jersiovic**
I'm doing this PR on behalf of @PinpointTownes replacing his original PR https://github.com/OrchardCMS/Orchard2/pull/603

### CODE REVIEW

File: src/Orchard.Cms.Web/Modules/Orchard.OpenId/Orchard.OpenId.csproj [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+  <PropertyGroup>
+    <VersionPrefix>2.0.0</VersionPrefix>
+    <TargetFramework>netstandard1.6</TargetFramework>
+    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <DebugType>portable</DebugType>
+    <AssemblyName>Orchard.OpenId</AssemblyName>
+    <PackageId>Orchard.OpenId</PackageId>
+    <NetStandardImplicitPackageVersion>1.6.1</NetStandardImplicitPackageVersion>
+    <GenerateAssemblyConfigurationAttribute>false</GenerateAssemblyConfigurationAttribute>
```
Comment (jersiovic) [inline src/Orchard.Cms.Web/Modules/Orchard.OpenId/Orchard.OpenId.csproj]: @PinpointTownes  Here I think you should add `<PackageTargetFallback>$(PackageTargetFallback);dnxcore50;portable-net451+win8</PackageTargetFallback>` like it is done in other .csproj projects?
Without this line the project is failing to me to compile.

File: src/Orchard.Cms.Web/Modules/Orchard.OpenId/Views/Admin/Edit.cshtml [hunk 2]
```diff
@@ -132,145 +118,147 @@
...
-    function refreshAllowRefreshTokenFlowVisibility() {
-        var allowRefreshTokenFlow = $("#AllowRefreshTokenFlow");
+        $("#AllowClientCredentialsFlow, #AllowPasswordFlow, #AllowAuthorizationCodeFlow, #AllowHybridFlow, #AllowImplicitFlow, #AllowRefreshTokenFlow").change(function () {
+            refreshFlows();
+        });
 
-        if (($("#AllowPasswordFlow").prop('checked') || $("#AllowAuthorizationCodeFlow").prop('checked') || $("#AllowHybridFlow").prop('checked'))) {
-            allowRefreshTokenFlow.removeAttr("disabled");
+        function refreshFlows() {
+            refreshRoleGroup();
```
Comment (jersiovic) [inline src/Orchard.Cms.Web/Modules/Orchard.OpenId/Views/Admin/Edit.cshtml]: You removed the ability of setting roles to a client app authenticated through Client Credentials, so this call and the called method should be removed.
However, the important thing is: why did you remove that feature? Did you enabled an alternate way of granting permissions to an app when Client Credentials is enabled?

File: src/Orchard.Cms.Web/Modules/Orchard.OpenId/Views/Admin/Edit.cshtml [hunk 3]
```diff
@@ -132,145 +118,147 @@
...
-        refreshAllowRefreshTokenFlowVisibility();
-        refreshRedirectSettings();
-    }
-    function refreshRoleGroup() {
-        $("#RoleGroup").collapse($("#AllowClientCredentialsFlow").prop('checked') ? "show" : "hide");
-    }
-    function refreshAllowRefreshTokenFlowVisibility() {
-        var allowRefreshTokenFlow = $("#AllowRefreshTokenFlow");
+        $("#AllowClientCredentialsFlow, #AllowPasswordFlow, #AllowAuthorizationCodeFlow, #AllowHybridFlow, #AllowImplicitFlow, #AllowRefreshTokenFlow").change(function () {
+            refreshFlows();
```
Comment (jersiovic) [inline src/Orchard.Cms.Web/Modules/Orchard.OpenId/Views/Admin/Edit.cshtml]: same here

File: src/Orchard.Cms.Web/Startup.cs [hunk 4]
```diff
@@ -36,6 +36,8 @@ public void ConfigureServices(IServiceCollection services)
 
             services.AddCommands();
 
+            services.AddAuthentication();
```
Comment (jersiovic) [inline src/Orchard.Cms.Web/Startup.cs]: Has this sth to do with the fact that in your original PR you skip changes in oauth branch for Orchard.Users/Startup.cs file? 
Those changes set dataProtectionProvider in IdentityOptions:
https://github.com/OrchardCMS/Orchard2/blob/oauth2/src/Orchard.Cms.Web/Modules/Orchard.Users/Startup.cs#L92
```
options.Cookies.ApplicationCookie.DataProtectionProvider = _dataProtectionProvider;
                options.Cookies.ExternalCookie.DataProtectionProvider = _dataProtectionProvider;
                options.Cookies.TwoFactorRememberMeCookie.DataProtectionProvider = _dataProtectionProvider;
                options.Cookies.TwoFactorUserIdCookie.DataProtectionProvider = _dataProtectionProvider;   
```
Should we undo those changes from oauth branch?

File: src/Orchard.Cms.Web/Startup.cs [hunk 5]
```diff
@@ -52,6 +54,7 @@ public void Configure(IApplicationBuilder app, IHostingEnvironment env, ILoggerF
             app.UseStaticFiles();
 
             loggerFactory.AddConsole(Configuration);
+            loggerFactory.AddDebug();
```
Comment (jersiovic) [inline src/Orchard.Cms.Web/Startup.cs]: I don't think this should be merged on master.

File: src/Orchard.Cms.Web/Modules/Orchard.OpenId/Orchard.OpenId.csproj (hunk 1 above)
Comment (PinpointTownes) [inline src/Orchard.Cms.Web/Modules/Orchard.OpenId/Orchard.OpenId.csproj]: All the packages used in this project target .NET Standard, so this shouldn't be necessary.

File: src/Orchard.Cms.Web/Modules/Orchard.OpenId/Views/Admin/Edit.cshtml (hunk 2 above)
Comment (PinpointTownes) [inline src/Orchard.Cms.Web/Modules/Orchard.OpenId/Views/Admin/Edit.cshtml]: Roles is a legacy concept in the new claims-based world, so everything should be done via claims. That said, if you want to preserve this feature, I'm fine.

File: src/Orchard.Cms.Web/Startup.cs (hunk 4 above)
Comment (PinpointTownes) [inline src/Orchard.Cms.Web/Startup.cs]: > @PinpointTownes Has this sth to do with the fact that in your original PR you skip changes in oauth branch for Orchard.Users/Startup.cs file?

No, it doesn't. For some reasons, Orchard no longer registers the authentication services by default: since the data protection provider is not available when initializing the modules, the OpenID module crashes because its `Startup` can't be instantiated.

> Should we undo those changes from oauth branch?

No. They should be directly added to the master branch to ensure tenants are correctly isolated. I'll open a separate issue to track that.

File: src/Orchard.Cms.Web/Startup.cs (hunk 5 above)
Comment (PinpointTownes) [inline src/Orchard.Cms.Web/Startup.cs]: It **should**. The console listener is useless when debugging an app hosted behind IIS (because well... there's no console).

File: src/Orchard.Cms.Web/Startup.cs (hunk 4 above)
Comment (PinpointTownes) [inline src/Orchard.Cms.Web/Startup.cs]: > I'll open a separate issue to track that.

Done: https://github.com/OrchardCMS/Orchard2/issues/606

File: src/Orchard.Cms.Web/Modules/Orchard.OpenId/Orchard.OpenId.csproj (hunk 1 above)
Comment (jersiovic) [inline src/Orchard.Cms.Web/Modules/Orchard.OpenId/Orchard.OpenId.csproj]: Do you mean it works for you withouth this line? Maybe is sth working wrong in my machine

File: src/Orchard.Cms.Web/Modules/Orchard.OpenId/Views/Admin/Edit.cshtml (hunk 2 above)
Comment (jersiovic) [inline src/Orchard.Cms.Web/Modules/Orchard.OpenId/Views/Admin/Edit.cshtml]: Roles is a legacy concept for authorization and it shouldn´t be used for authorize but it doesn't means it is used for grouping claims. As actually Orchard uses it. I don't think we need to inform of roles in access token, what we need is to provide claims related to roles assigned to an app. Like we did in oauth branch here:
https://github.com/OrchardCMS/Orchard2/blob/oauth2/src/Orchard.Cms.Web/Modules/Orchard.OpenId/Services/OpenIdUserManager.cs#L37
and here: https://github.com/OrchardCMS/Orchard2/blob/oauth2/src/Orchard.Cms.Web/Modules/Orchard.OpenId/Services/OpenIdApplicationManager.cs#L78
So, if do you agree I will undo your changes on UI for setting roles for client credentials flows and I will add here https://github.com/OrchardCMS/Orchard2/blob/openidUpgrade/src/Orchard.Cms.Web/Modules/Orchard.OpenId/Controllers/AccessController.cs#L334 code for adding claims related with roles assigned to an app.

File: src/Orchard.Cms.Web/Startup.cs (hunk 4 above)
Comment (jersiovic) [inline src/Orchard.Cms.Web/Startup.cs]: ok, if oauth branch is finally merged to master this issue also will be fixed.

File: src/Orchard.Cms.Web/Startup.cs (hunk 5 above)
Comment (jersiovic) [inline src/Orchard.Cms.Web/Startup.cs]: ah ok, I see your point

File: src/Orchard.Cms.Web/Modules/Orchard.OpenId/Orchard.OpenId.csproj (hunk 1 above)
Comment (PinpointTownes) [inline src/Orchard.Cms.Web/Modules/Orchard.OpenId/Orchard.OpenId.csproj]: Yup, it works fine on my machine. I took a look at the CI build logs and the OpenID module compiles fine on AppVeyor. That said, if you really need this line to make it work, feel free to add it... it's harmless.

File: src/Orchard.Cms.Web/Modules/Orchard.OpenId/Views/Admin/Edit.cshtml (hunk 2 above)
Comment (PinpointTownes) [inline src/Orchard.Cms.Web/Modules/Orchard.OpenId/Views/Admin/Edit.cshtml]: > So, if do you agree I will undo your changes on UI for setting roles for client credentials flows and I will add here 

I'm fine with that.

File: src/Orchard.Cms.Web/Modules/Orchard.OpenId/Controllers/AdminController.cs [hunk 6]
```diff
@@ -223,11 +189,20 @@ public class AdminController : Controller
                 AllowHybridFlow = model.AllowHybridFlow
             };
 
-            await _applicationManager.CreateAsync(openIdApp);
+            if (model.Type == ClientType.Confidential)
+            {
+                await _applicationManager.CreateAsync(application, model.ClientSecret, HttpContext.RequestAborted);
```
Comment (jersiovic) [inline src/Orchard.Cms.Web/Modules/Orchard.OpenId/Controllers/AdminController.cs]: This line throws an exception if you create a new confidential app:
{System.ArgumentException: Requested value 'confidential' was not found.
   at System.Enum.TryParseEnum(Type enumType, String value, Boolean ignoreCase, EnumResult& parseResult)
   at System.Enum.Parse(Type enumType, String value, Boolean ignoreCase)
   at Orchard.OpenId.Services.OpenIdApplicationStore.SetClientTypeAsync(OpenIdApplication application, String type, CancellationToken cancellationToken) in D:\PROYECTOS\OrchardCMS\Orchard2\src\Orchard.Cms.Web\Modules\Orchard.OpenId\Services\OpenIdApplicationStore.cs:line 154
   at OpenIddict.Core.OpenIddictApplicationManager`1.<CreateAsync>d__8.MoveNext()}
Problem happens because public Task SetClientTypeAsync(OpenIdApplication application, string type, CancellationToken cancellationToken) is called with confidential type in lowercase despite Type OpenIdApplication is an enum.

I though it could be caused by this ToLower()
https://github.com/OrchardCMS/Orchard2/blob/openidUpgrade/src/Orchard.Cms.Web/Modules/Orchard.OpenId/Services/OpenIdApplicationStore.cs#L65
 but after removing it, it still happens that type received in SetClientTypeAsync is lowercase and this causes an exception
It looks openiddict provides the type in lowercase, so we have to patch it changing first letter to upper case.

File: src/Orchard.Cms.Web/Modules/Orchard.OpenId/Controllers/AdminController.cs [hunk 7]
```diff
@@ -134,31 +128,17 @@ public class AdminController : Controller
...
-            app.AllowAuthorizationCodeFlow = model.AllowAuthorizationCodeFlow;
-            app.AllowClientCredentialsFlow = model.AllowClientCredentialsFlow;
-            app.AllowImplicitFlow = model.AllowImplicitFlow;
-            app.AllowPasswordFlow = model.AllowPasswordFlow;
-            app.AllowRefreshTokenFlow = model.AllowRefreshTokenFlow;
-            app.AllowHybridFlow = model.AllowHybridFlow;
-
-            await _applicationManager.CreateAsync(app);
+            await TryUpdateModelAsync(application);
+            await _applicationManager.UpdateAsync(application, HttpContext.RequestAborted);
```
Comment (jersiovic) [inline src/Orchard.Cms.Web/Modules/Orchard.OpenId/Controllers/AdminController.cs]: If you create a Public application and in a second step you edit it changing its type to Confidential, you get an exception cause a password is not provided.

File: src/Orchard.Cms.Web/Modules/Orchard.OpenId/Controllers/AdminController.cs (hunk 6 above)
Comment (jersiovic) [inline src/Orchard.Cms.Web/Modules/Orchard.OpenId/Controllers/AdminController.cs]: In the commit I'm working on based on what we comment here I've added a dirty check that works. 
```
public Task SetClientTypeAsync(OpenIdApplication application, string type, CancellationToken cancellationToken)
        {
            if (application == null)
            {
                throw new ArgumentNullException(nameof(application));
            }

            if (type == null || type.Length<1)
            {
                throw new ArgumentNullException(nameof(type));
            }

            type= type[0].ToString().ToUpper() + type.Substring(1);
            application.Type = (ClientType) Enum.Parse(typeof(ClientType), type);

            return Task.CompletedTask;
        }
```

File: src/Orchard.Cms.Web/Modules/Orchard.OpenId/Controllers/AdminController.cs (hunk 7 above)
Comment (jersiovic) [inline src/Orchard.Cms.Web/Modules/Orchard.OpenId/Controllers/AdminController.cs]: I will file an issue with that, to provide also a way of setting a client secret for an existing app.

File: src/Orchard.Cms.Web/Modules/Orchard.OpenId/Controllers/AdminController.cs (hunk 6 above)
Comment (PinpointTownes) [inline src/Orchard.Cms.Web/Modules/Orchard.OpenId/Controllers/AdminController.cs]: Hum, good point. I guess we could use the `Enum.Parse` overload accepting an `ignoreCase` parameter.

File: src/Orchard.Cms.Web/Modules/Orchard.OpenId/Controllers/AdminController.cs (hunk 7 above)
Comment (PinpointTownes) [inline src/Orchard.Cms.Web/Modules/Orchard.OpenId/Controllers/AdminController.cs]: Yeah. We'll likely need to update OpenIddict to expose an `UpdateAsync` overload accepting the client secret parameter.

File: src/Orchard.Cms.Web/Startup.cs (hunk 5 above)
Comment (sebastienros) [inline src/Orchard.Cms.Web/Startup.cs]: Should it should at least be registered only on Development environment.

File: src/Orchard.Cms.Web/Modules/Orchard.Roles/Services/RoleStore.cs [hunk 8]
```diff
@@ -58,10 +58,10 @@ public void UpdateRoles(RolesDocument roles)
             _memoryCache.Set(Key, roles);
         }
 
-        public async Task<IEnumerable<string>> GetRoleNamesAsync()
+        public async Task<IEnumerable<RoleNamesEntry>> GetRoleNamesAsync()
```
Comment (jersiovic) [inline src/Orchard.Cms.Web/Modules/Orchard.Roles/Services/RoleStore.cs]: @sebastienros I did this change because RoleStore only allows to retrieve a NormalizedName, but not by Name. But I need both, Name for showing it to the user and the other to be able of retrieving a Role per NormalizedName to get its Claims in AccessController

File: src/Orchard.Cms.Web/Modules/Orchard.OpenId/Views/Admin/Edit.cshtml [hunk 9]
```diff
@@ -28,6 +28,26 @@
         </select>
         <div class="hint">@T["Confidential applications MUST send their client secret when communicating with the token and revocation endpoints. This guarantees that only the legit client can exchange an authorization code or get a refresh token."]</div>
     </div>
+    <fieldset class="form-group">
+        <div class="form-check">
```
Comment (jersiovic) [inline src/Orchard.Cms.Web/Modules/Orchard.OpenId/Views/Admin/Edit.cshtml]: Finally I added a way for updating Client Secret when an application is edited

File: src/Orchard.Cms.Web/Modules/Orchard.Roles/Controllers/AdminController.cs [hunk 10]
```diff
@@ -72,7 +72,7 @@ IEnumerable<IPermissionProvider> permissionProviders
 
             var model = new RolesViewModel
             {
-                RoleEntries = roles.Select(BuildRoleEntry).ToList()
+                RoleEntries = roles.Select(r => BuildRoleEntry(r.Name)).ToList()
```
Comment (sebastienros) [inline src/Orchard.Cms.Web/Modules/Orchard.Roles/Controllers/AdminController.cs]: Useless change

File: src/Orchard.Cms.Web/Modules/Orchard.Users/Controllers/AdminController.cs [hunk 11]
```diff
@@ -303,7 +303,7 @@ IUserService userService
 
         public async Task<IEnumerable<string>> GetRoleNamesAsync()
         {
-            var roleNames = await _roleProvider.GetRoleNamesAsync();
+            var roleNames = (await _roleProvider.GetRoleNamesAsync()).Select(r=>r.Name);
```
Comment (sebastienros) [inline src/Orchard.Cms.Web/Modules/Orchard.Users/Controllers/AdminController.cs]: `ToArray` or it will be applied multiple times

File: src/Orchard.Cms.Web/Startup.cs [hunk 12]
```diff
@@ -53,8 +53,11 @@ public void Configure(IApplicationBuilder app, IHostingEnvironment env, ILoggerF
 
             app.UseStaticFiles();
 
-            loggerFactory.AddConsole(Configuration);
-            loggerFactory.AddDebug();
+            if (env.IsDevelopment())
```
Comment (sebastienros) [inline src/Orchard.Cms.Web/Startup.cs]: Only the `Debug` should be in Development, the prod might use the console. And the console can be driven by configuration, unlike debug.

File: src/Orchard.Cms.Web/Modules/Orchard.OpenId/Controllers/AdminController.cs [hunk 13]
```diff
@@ -127,12 +153,15 @@ public class AdminController : Controller
...
 
-            var application = await _applicationManager.FindByIdAsync(model.Id, HttpContext.RequestAborted);
-            if (application == null)
-                return NotFound();
+            application.NormalizedRoleNames = new List<string>();
+            if (application.Type == ClientType.Confidential && application.AllowClientCredentialsFlow)
+                application.NormalizedRoleNames = model.RoleEntries.Where(r => r.Selected).Select(r => r.NormalizedName).ToList();
 
-            await TryUpdateModelAsync(application);
             await _applicationManager.UpdateAsync(application, HttpContext.RequestAborted);
```
Comment (jersiovic) [inline src/Orchard.Cms.Web/Modules/Orchard.OpenId/Controllers/AdminController.cs]: I have tried to avoid problem of UpdateAsync method that throws an exception if you attempt to modify a confidential app because it doesn't have a client secret calling to await _applicationManager.SetClientSecretAsync() at line 159, but this also returns an exception. So we need a fix on Openiddict.
I left the code with this attempt because no matter if it is there or not the exception will always happen.

File: src/Orchard.Cms.Web/Modules/Orchard.Roles/Services/RoleStore.cs (hunk 8 above)
Comment (sebastienros) [inline src/Orchard.Cms.Web/Modules/Orchard.Roles/Services/RoleStore.cs]: What if you need to "display" something else next?

Can you remind me where `IRoleProvider` is used, and if it's our code or AspNet one?

**Comment by jersiovic** [comment]
Please review if you agree with my changes to be merged in oauth branch. After alpha2 version of openiddict was removed from nuget I cannot build the solution for my customers based on current bits in oauth2, so I'm a bit in a hurry in order to propagate this changes to my fork.

File: src/Orchard.Cms.Web/Modules/Orchard.Roles/Services/RoleStore.cs (hunk 8 above)
Comment (jersiovic) [inline src/Orchard.Cms.Web/Modules/Orchard.Roles/Services/RoleStore.cs]: All the code I saw was our code. This is a sample of use of IRoleProvider https://github.com/OrchardCMS/Orchard2/blob/openidUpgrade/src/Orchard.Cms.Web/Modules/Orchard.OpenId/Controllers/AdminController.cs#L107
The point is I need to store NormalizedName  here https://github.com/OrchardCMS/Orchard2/blob/openidUpgrade/src/Orchard.Cms.Web/Modules/Orchard.OpenId/Models/OpenIdApplication.cs#L63

Because I need to know the Normalized name for retrieving claims https://github.com/OrchardCMS/Orchard2/blob/openidUpgrade/src/Orchard.Cms.Web/Modules/Orchard.OpenId/Controllers/AccessController.cs#L345

File: src/Orchard.Cms.Web/Modules/Orchard.Roles/Controllers/AdminController.cs (hunk 10 above)
Comment (jersiovic) [inline src/Orchard.Cms.Web/Modules/Orchard.Roles/Controllers/AdminController.cs]: So, do I change BuildRoleEntry method to expect a RoleNamesEntry type instead of string as parameter? 
Or do you prefer IRoleProvider provides another method that returns only names?
Of directly do we provide a method returning roles and each consumer will take what it needs?

File: src/Orchard.Cms.Web/Modules/Orchard.Users/Controllers/AdminController.cs (hunk 11 above)
Comment (jersiovic) [inline src/Orchard.Cms.Web/Modules/Orchard.Users/Controllers/AdminController.cs]: Ups!

File: src/Orchard.Cms.Web/Startup.cs (hunk 12 above)
Comment (jersiovic) [inline src/Orchard.Cms.Web/Startup.cs]: ok

File: src/Orchard.Cms.Web/Modules/Orchard.Roles/Services/RoleStore.cs (hunk 8 above)
Comment (sebastienros) [inline src/Orchard.Cms.Web/Modules/Orchard.Roles/Services/RoleStore.cs]: Why not return the normalized name only then, and if you need the displayed one then you just look it up using the service?

File: src/Orchard.Cms.Web/Modules/Orchard.Roles/Controllers/AdminController.cs (hunk 10 above)
Comment (sebastienros) [inline src/Orchard.Cms.Web/Modules/Orchard.Roles/Controllers/AdminController.cs]: I understand this change now, didn't see the Entry before. Ignore this comment.

File: src/Orchard.Cms.Web/Modules/Orchard.Roles/Services/RoleStore.cs (hunk 8 above)
Comment (jersiovic) [inline src/Orchard.Cms.Web/Modules/Orchard.Roles/Services/RoleStore.cs]: It is less optimal because we will read roles document for each role but it makes sense and performance in this scenario is not critical. I will do it.

**Comment by jersiovic** [comment]
Done. 
Finally I noticed in previous commit I get confused by _roleStore.FindByIdAsync() and _roleStore.FindByNameAsync() methods. Their name doesn't match too much with what they do.
First one looks by Name and second one looks by NormalizedName.
So, with _roleStore.FindByIdAsync() we don't need to use NormalizedName everywhere.

File: src/Orchard.Cms.Web/Modules/Orchard.OpenId/Controllers/AdminController.cs (hunk 7 above)
Comment (jersiovic) [inline src/Orchard.Cms.Web/Modules/Orchard.OpenId/Controllers/AdminController.cs]: I have filed an issue with that https://github.com/OrchardCMS/Orchard2/issues/609

**Comment by PinpointTownes** [comment]
@sebastienros can we consider merging the `oauth2` into `master` now that the module is quite stable?

**Comment by jersiovic** [comment]
I was waiting the following for merging to master:
- To check if in my preproduction server the app updated with latest bits of oauth branch works properly.
- This PR is approved https://github.com/OrchardCMS/Orchard2/pull/563
- This issue is fixed https://github.com/OrchardCMS/Orchard2/issues/609 
But if you don't mind we fix later on main branch posible problems we find or we merge later those PR I don't have any inconvenience on creating a PR to merge to master.

**Comment by PinpointTownes** [comment]
@jersiovic @sebastienros now that https://github.com/OrchardCMS/Orchard2/pull/615 was merged, maybe it's the perfect time to do that?

**Comment by sebastienros** [comment]
LGTM

**Comment by sebastienros** [comment]
I'll let @jersiovic do it as there might be some other things to do and ensure it works. But you have my approval to do it when you want.
